src/Bicep.LangServer/Handlers/InsertResourceHandler.cs

```diff
@@ -0,0 +1,298 @@
+                    case "type":
+                    case "apiVersion":
+                        // Don't add these to the resource properties - they're part of the resource declaration.
+                        break;
+                    case "name":
+                        // Use the fully-qualified name instead of the name returned by the RP.
+                        properties.Add(SyntaxFactory.CreateObjectProperty(
+                            "name",
+                            SyntaxFactory.CreateStringLiteral(resourceId.FullyQualifiedName)));
+                        break;
+                    default:
+                        properties.Add(SyntaxFactory.CreateObjectProperty(
+                            property.Name,
+                            ConvertJsonElement(property.Value)));
+                        break;
+                }
+            }
+
+            var description = SyntaxFactory.CreateDecorator(
+                "description",
+                SyntaxFactory.CreateStringLiteral($"Generated from {resourceId.FullyQualifiedId}"));
+
+            return new ResourceDeclarationSyntax(
+                new SyntaxBase[] { description, SyntaxFactory.NewlineToken, },
+                SyntaxFactory.CreateToken(TokenType.Identifier, "resource"),
+                SyntaxFactory.CreateIdentifier(Regex.Replace(resourceId.UnqualifiedName, "[^a-zA-Z]", "")),
+                SyntaxFactory.CreateStringLiteral(typeReference.FormatName()),
+                null,
+                SyntaxFactory.CreateToken(TokenType.Assignment),
+                SyntaxFactory.CreateObject(properties));
+        }
+
+        private static IAzResourceProvider.AzResourceIdentifier? TryParseResourceId(string? resourceIdString)
+        {
+            if (resourceIdString is null)
+            {
+                return null;
+            }
+
+            if (ResourceId.TryParse(resourceIdString, out var resourceId))
+            {
+                return new(
+                    resourceId.FullyQualifiedId,
+                    resourceId.FormatFullyQualifiedType(),
+                    resourceId.FormatName(),
+                    resourceId.NameHierarchy.Last());
+            }
+
+            var rgRegexOptions = RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture | RegexOptions.Compiled | RegexOptions.CultureInvariant;
+            var rgRegex = new Regex(@"^/subscriptions/(?<subId>[^/]+)/resourceGroups/(?<rgName>[^/]+)$", rgRegexOptions);
+            var rgRegexMatch = rgRegex.Match(resourceIdString);
+            if (rgRegexMatch.Success)
+            {
+                return new(
+                    resourceIdString,
+                    "Microsoft.Resources/resourceGroups",
+                    rgRegexMatch.Groups["rgName"].Value,
+                    rgRegexMatch.Groups["rgName"].Value);
+            }
+
+            return null;
+        }
+
+        private static SyntaxBase ConvertJsonElement(JsonElement element)
```

Comment:
>ConvertJsonElement [](http://example.com/codeflow?start=34&length=18)

Could we reuse the decompiler and save ourselves some code? Here's the flow I'm thinking:
1. GET the resource
2. Wrap the resource in a template JSON (should be a pretty simple transformation)
3. Decompile
4. Insert the resource declaration that got decompiled

Reply:
I did consider this, but there are some differences, and in practice, the following 30 lines are really all the logic I would need to share - it felt better to duplicate.

1. Here we're using System.Text.Json - the decompiler uses Newtonsoft.
1. We need to invoke a different set of rewriters, and want to use a different symbolic naming convetion.
1. None of the expression rewriting logic is required.

Reply:
Makes sense.